infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/RuntimeCleaner.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes;
+
+import static org.eclipse.che.workspace.infrastructure.kubernetes.server.external.SingleHostExternalServiceExposureStrategy.SINGLE_HOST_STRATEGY;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.CheNamespace;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespace;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.WorkspaceExposureType;
+
+public class RuntimeCleaner {
+
+  private final boolean cleanupCheNamespace;
+
+  private final CheNamespace cheNamespace;
+
+  @Inject
+  public RuntimeCleaner(
+      @Named("che.infra.kubernetes.server_strategy") String exposureStrategy,
+      @Named("che.infra.kubernetes.singlehost.workspace.exposure") String singleHostStrategy,
+      CheNamespace cheNamespace) {
+    this.cheNamespace = cheNamespace;
+
+    this.cleanupCheNamespace =
```

Comment:
Corner case - This will not clean up the workspace that started in singlehost mode and the server got redeployed in multihost without stopping the workspace first.

Reply:
I know https://github.com/eclipse/che/issues/17999#issuecomment-702237320 :) Any ideas how to fix it?

Reply:
Just ignore it. ))
Does the restart of the workspace will fix the situation?

Reply:
or you can check if there any running workspaces and if yes just fail to start or print warning

Reply:
Workspaces will work, just configmaps with traefik configs will stay in che namespaces.

Reply:
Are they going to be cleaned on the next workspace restart?

Reply:
no, because the che workspace cleanup is behind if, which will get execute only on gateway single-host
```
if (cleanupCheNamespace) {
      cheNamespace.cleanUp(workspaceId);
    }
```

Reply:
I don't think that is critical